### PR TITLE
SparseSeries accepts scipy.sparse.spmatrix in constructor

### LINKED ISSUE #15634: API/DEPR: deprecate SparseSeries.from_coo and accept in constructor
xref #15497 

using ``SparseDataFrame`` as the model.

### PR DESCRIPTION
**Pull request by kernc**
- [X] closes #15634
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [X] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.20.2.txt [hunk 1]
```diff
@@ -25,6 +25,9 @@ Enhancements
   has been added to return the group order (:issue:`11642`); see
   :ref:`here <groupby.ngroup>`.
 
+
+- ``SparseSeries`` and ``SparseArray`` now support 1d ``scipy.sparse.spmatrix`` in constructor. Additionally, ``SparseDataFrame`` can be assigned columns of ``scipy.sparse.spmatrix``; see :ref:`here <sparse.scipysparse_series>`. (:issue:`15634`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: will be for 0.21.0

File: pandas/core/sparse/frame.py [hunk 2]
```diff
@@ -433,6 +433,15 @@ def __getitem__(self, key):
...
             return self._get_item_cache(key)
 
+    def __setitem__(self, key, value):
+        if is_scipy_sparse(value):
+            if any(ax == 1 for ax in value.shape):  # 1d spmatrix
+                value = SparseArray(value, fill_value=self._default_fill_value)
+            else:
+                # 2d; make it iterable
+                value = list(value.tocsc().T)
+        super().__setitem__(key, value)
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: use the fully qualified call

File: pandas/core/sparse/series.py [hunk 3]
```diff
@@ -722,6 +726,9 @@ def combine_first(self, other):
 
     def to_coo(self, row_levels=(0, ), column_levels=(1, ), sort_labels=False):
         """
+        DEPRECATED; instead, make a SparseSeries with a two-level index,
+        unstack it, then use .to_coo() on the resulting SparseDataFrame.
```
Comment (jreback) [inline pandas/core/sparse/series.py]: use the deprecated sphinx directive (I think we are changing these all over)

File: pandas/core/sparse/series.py [hunk 4]
```diff
@@ -779,6 +786,9 @@ def to_coo(self, row_levels=(0, ), column_levels=(1, ), sort_labels=False):
     @classmethod
     def from_coo(cls, A, dense_index=False):
         """
+        DEPRECATED; instead, pass 1d scipy.sparse matrices directly into
+        SparseSeries constructor, and 2d into SparseDataFrame constructor.
```
Comment (jreback) [inline pandas/core/sparse/series.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]
just a quick look. this is probably going to be tricky to make work.

**Comment by kernc** [comment]
> this is probably going to be tricky to make work.

Why do you think so? I does seem to work for the moment. 😃

File: doc/source/sparse.rst [hunk 5]
```diff
@@ -213,9 +213,28 @@ To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you
...
+
+   sa = pd.SparseSeries(sp_arr[:, 5])
+   sa
+
+   sdf['x'] = sa
+   sdf['y'] = sp_arr[:, 6]
+   sdf[['z', 'w']] = sp_arr[:, [7, 8]]
+   sdf.iloc[:, -5:]
+
+Below interface is deprecated.
```
Comment (jreback) [inline doc/source/sparse.rst]: say that this is deprecated in 0.21.0

File: doc/source/whatsnew/v0.21.0.txt [hunk 6]
```diff
@@ -25,6 +25,10 @@ New features
 - Added `__fspath__` method to :class`:pandas.HDFStore`, :class:`pandas.ExcelFile`,
   and :class:`pandas.ExcelWriter` to work properly with the file system path protocol (:issue:`13823`)
 
+- ``SparseSeries`` and ``SparseArray`` now support 1d ``scipy.sparse.spmatrix`` in constructor.
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: in the constructor

File: doc/source/whatsnew/v0.21.0.txt [hunk 7]
```diff
@@ -25,6 +25,10 @@ New features
...
+  Additionally, ``SparseDataFrame`` can be assigned columns of ``scipy.sparse.spmatrix``;
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: make this 2nd sentence a separate bullet point (you can use same issue on both of them, or 2nd one should be the PR number maybe)

File: pandas/core/sparse/frame.py [hunk 8]
```diff
@@ -433,6 +433,16 @@ def __getitem__(self, key):
...
             return self._get_item_cache(key)
 
+    def __setitem__(self, key, value):
+        if is_scipy_sparse(value):
+            if any(ax == 1 for ax in value.shape):  # 1d spmatrix
+                value = SparseArray(value, fill_value=self._default_fill_value,
+                                    kind=self._default_kind)
+            else:
+                # 2d; make it iterable
+                value = list(value.tocsc().T)
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: does this materialize?

File: pandas/tests/sparse/test_frame.py [hunk 9]
```diff
@@ -540,6 +540,37 @@ def test_setitem_array(self):
...
+        sdf = self.frame.copy(False)
+
+        def _equal(spm1, spm2):
+            return np.all(spm1.toarray() == spm2.toarray())
+
+        # 1d -- column
+        spm = csr_matrix(np.arange(len(sdf))).T
+        sdf['X'] = spm
+        assert _equal(sdf[['X']].to_coo(), spm)
+
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: this comparision on the scipy side is fine, but also let's compare with assert_sparse_series/frame_equal

File: pandas/tests/sparse/test_frame.py [hunk 10]
```diff
@@ -540,6 +540,37 @@ def test_setitem_array(self):
...
+        # 1d -- existing column
+        sdf['A'] = spm.T
+        assert _equal(sdf[['X']].to_coo(), spm)
+
+        # 1d row -- changing series contents not yet supported
+        spm = csr_matrix(np.arange(sdf.shape[1], dtype=float))
+        idx = np.zeros(sdf.shape[0], dtype=bool)
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: can you test with ``.loc/.iloc`` as well (might already be another issue about this, if not and its too complicated for here, then create a new issue)

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good. a couple of comments.

**Comment by jreback** [comment]
can you rebase / update

**Comment by jreback** [comment]
@kernc if you have time can you rebase / update

**Comment by kernc** [comment]
For some reason (probably `assert_raises` in `test_setitem_spmatrix()`), I had a while ago decided to first implement `setitem` on `SparseSeries` that I have 80% done waiting in a feature branch somewhere. If you insist, I can push this first, however.

**Comment by jreback** [comment]
oh we can do the other first 
was just looking thru open prs 

what was that number?

**Comment by kernc** [comment]
Right, it's https://github.com/pandas-dev/pandas/pull/17785.

**Comment by jreback** [comment]
closing as stale. if you want to continue working, pls ping.

**Pull request closed without merging**